src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+for handling different events on the widget.  There's also a very nice
+`tutorial <https://facebook.github.io/react/tutorial/tutorial.html>`_ for
+building a tic-tac-toe game.  If you can grok that then you'll have the hang of
+React.
+
+If you've ever used a GUI toolkit like wx or Swing it shouldn't be too hard
+to pick up on what it's doing.
+
+Using React is quite a bit different from the old-fashioned way of making
+reactive web UIs with JavaScript.  What I'm calling the "old-fashioned" way is
+a couple things.  For one, the server might render serve up a bunch of HTML
+containing all the elements in your page, many of which might by "hidden" using
+CSS, and the JavaScript would hide and unhide elements on the page.  Or the
+JavaScript might generate some elements and insert them directly into the DOM
+and remove them as needed, either using the DOM API directly or, somewhat
+later, tools like `jQuery`_ (note: jQuery still has a role to play even in
+conjunction with React though).
+
+In other words, gone are the days of servers rendering and returning HTML to
+the browser.  All the rendering is pushed entirely to the client, with the
+client-server communication focused on as light-weight as possible message
+passing.  This potentially frees up enormous resources for the server, while
+pushing much more work to the client (which is why so many of your browser
+tabs are using over 100 MB of memory, among other reasons).
+
+The way React works, in short, is this: It maintains its own "virtual DOM"
+separate from the actual DOM of the browser document, with the same API as the
+real DOM.  Whenever you show, hide, or otherwise update the contents of a UI
+element in the application, it uses a copy of its virtual DOM to figure out
+exactly what needs to change in order for that to happen, and generates (and
+subsequently applies) a stream of operations to perform on the actual DOM in
+order to enact those changes.  The result is that there's nothing in the real
+DOM except for what's actually displayed on the page, which is convenient for
+debugging and inspection via your browser's development tools.  There's a
+simple `demonstration
+<https://facebook.github.io/react/docs/rendering-elements.html#react-only-updates-whats-necessary>`_
+of this aspect in the docs.
+
+Another nice aspect of React is its JSX domain-specific language which I'll
+discuss more next.
+
+As mentioned in the section on pug, essentially all of SMC's web frontend is
+built using React.  Almost no HTML is ever sent from the server.  Instead
+the frontend is built up by React.  When user interactions with the UI need
+to be persisted, those are sent as event messages (typically over
+WebSockets) to the server, which may in turn respond with events that result
+in updating the UI in some appropriate way (the event messages are usually a
+JSON object of some kind).  This is still an over-simplification (see for
+example the section on Redux later), but that's the basic idea.
+
```

Comment:
>  In order to do much beyond that, codemirror would have to be rewritten in React as well...

Which is pointless.  It would presumably be identical from the outside as a black box (same API), so why write it "in react".    For comparison, Github's Atom.io editor *was* written in React for a while (with the same outside api), but then they rewrote it directly in Javascript to get a speedup.  With very careful attention and thought, direct Javascript will always be better than React, and Codemirror has already got years and  years of that careful attention, thought, and optimization.  (I think codemirror is significantly more optimized than atom.io by the way.)